### ISSUE TITLE
How can I wrap sonnet module nicely with tf.function?

### ISSUE THREAD

**Issue by mingyr** [deepmind/sonnet#201]
Dear Deepminder,

Recently I am updating my code from Sonnet 1.x to 2.x. I have a question.

I still want to keep the graph aspect of TensorFlow code, so I prefer to wrap the sonnet module with tf.function to speed the training process up.

The difficulty lies here, referring to the following code:

```
    ....
    for it in range(start_iteration, FLAGS.num_iterations):
        with tf.GradientTape(persistent = True) as tape:
            images, labels = input_(os.path.join(FLAGS.data_dir, 'train.tfr'))
            logits = model(images)
            loss = LossClassification(FLAGS.num_classes)(logits, tf.squeeze(labels))

        opt(loss, model.trainable_variables, tape = tape)
    ....
```
 
if I do in the following way:

`           logits = tf.function(model)(images)`

Then the following reference of model.trainable_variables in _opt(loss, model.trainable_variables, tape = tape)_ would complain that tf.function object doesn't have the property trainable_variables.

Since the property trainable_variables of sonnet model is only valid if the module is connected, another way might be:

```
            images, labels = input_(os.path.join(FLAGS.data_dir, 'train.tfr'))
            logits = model(images)
            trainable_variables = model.trainable_variables
            logits = tf.function(model)(images)
            loss = LossClassification(FLAGS.num_classes)(logits, tf.squeeze(labels))
```

But this is awkward, isn't it? since _logits_ will be calculated twice. So could I hear your suggestoin in this regard?

**Comment by tomhennigan** [comment]
Hi @mingyr, in this case I would suggest having two variables, one referencing the model and the other being a tf.function applying it:

```python
model = snt.nets.MLP([300, 100, 10])
infer = tf.function(model)
logits = infer(x)
params = model.trainable_variables
```

In general you want to put as much code as you can inside a `tf.function`, so I would suggest looking at ways to put both the logits computation and the loss computation, gradient computation and optimizer update into a single `tf.function`. We have an example of this in our [MNIST example](https://github.com/deepmind/sonnet/blob/v2/examples/simple_mnist.py) and it should be straightforward to make that data parallel over many GPUs following our [distributed Cifar10](https://github.com/deepmind/sonnet/blob/v2/examples/distributed_cifar10.ipynb) example.

**Comment by mingyr** [comment]
Hey, Tom, yes, you are right, appreciate your quick and clear explanation.
